teps/0041-move-image-entrypoint-lookup-to-the-taskrun-pod.md

```diff
@@ -0,0 +1,161 @@
+---
+status: proposed
+title: Move Image Entrypoint Lookup to the TaskRun Pod
+creation-date: '2021-01-11'
+last-updated: '2021-01-21'
+authors:
+- '@yaoxiaoqi'
+---
+
+# TEP-0041: Move Image Entrypoint Lookup to the TaskRun Pod
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+- [Proposal](#proposal)
+  - [Add Flags to EntryPoint Binary](#add-flags-to-entrypoint-binary)
+  - [Resolve Image in EntryPoint Binary](#resolve-image-in-entrypoint-binary)
+  - [Notes/Caveats](#notescaveats)
+    - [Entrypoint Cache](#entrypoint-cache)
+    - [Error Reporting Delay](#error-reporting-delay)
+    - [Inconsistency](#inconsistency)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes to move the image entrypoint lookup to entrypoint binary that
+runs in the TaskRun pod.
+
+## Motivation
+
+Currently, Tekton controller pod has access to all service accounts and will do
+entry point lookup using `go-containerregistry` for the images that don't have
+any command specified for them. This potentially might cause permission denied
+error when users are trying to pull private images from registry since the
+default service account and the service account that Tekton controller is using
+can't authenticate the requests on some of cloud platforms. The private image is
+actually running on the TaskRun pod so this will involve some inter-pod
+permission issue.
+
+Therefore, this TEP proposes to move the image entrypoint lookup to TaskRun pod.
+By doing this, we can use the service account that user provided for the TaskRun
+to query the entrypoint of the image that is running on this pod. This will also
+avoid the authentication problem when requests come from different pods on some
+cloud providers and the confusing retricted access to the image when command is not
+specified.
+
+### Goals
+
+- Move the image entry point lookup logic altogether to the entry point binary on the TaskRun pod
+
+### Non-Goals
+
+- Implement a kubernetes api to query the metadata of a image using kubelet
+
+### Use Cases
+
+TaskRun users encounter permission denied error if they don't specify command in
+their container specs on some cloud providers. The error happens when TaskRun
```

Comment:
To play devil's advocate here - poking at the code it looks like `go-containerregistry` only looks at [`imagePullSecrets`](https://github.com/google/go-containerregistry/blob/c95d7bdd6359dab9cbf575fae14e3389e3efa84e/pkg/authn/k8schain/k8schain.go#L71-L91) before defaulting to [anonymous fetching](https://github.com/google/go-containerregistry/blob/c95d7bdd6359dab9cbf575fae14e3389e3efa84e/pkg/authn/k8schain/k8schain.go#L171) (this is where any provider specific identity tied to the SA would come into play).

Would we be able to provide an [`authn.Keychain`](https://pkg.go.dev/github.com/google/go-containerregistry@v0.3.0/pkg/authn#Keychain) that would replicate the behavior we want? (and what extra permissions would be needed if any?)

Reply:
cc @mattmoor